sm/agent/src/agent.rs

```diff
@@ -0,0 +1,326 @@
+
+        Self {
+            config: SmAgentConfig::default(),
+            name: name.into(),
+            user_manager,
+            config_location,
+            persistance_store,
+        }
+    }
+
+    pub async fn start(&self) -> Result<(), AgentError> {
+        info!("Starting sm-agent");
+
+        let plugins = Arc::new(ExternalPlugins::open("/etc/tedge/sm-plugins")?);
+        if plugins.empty() {
+            error!("Couldn't load plugins from /etc/tedge/sm-plugins");
+            return Err(AgentError::NoPlugins);
+        }
+
+        let mqtt = Client::connect(self.name.as_str(), &self.config.mqtt_client_config).await?;
+        let mut errors = mqtt.subscribe_errors();
+        tokio::spawn(async move {
+            while let Some(error) = errors.next().await {
+                error!("{}", error);
+            }
+        });
+
+        let () = self.check_state_store(&mqtt).await?;
+
+        // * Maybe it would be nice if mapper/registry responds
+        let () = publish_capabilities(&mqtt).await?;
+
+        let () = self.subscribe_and_process(&mqtt, &plugins).await?;
+
+        Ok(())
+    }
+
+    async fn subscribe_and_process(
+        &self,
+        mqtt: &Client,
+        plugins: &Arc<ExternalPlugins>,
+    ) -> Result<(), AgentError> {
+        let mut operations = mqtt.subscribe(self.config.request_topics.clone()).await?;
+        while let Some(message) = operations.next().await {
+            info!("Request {:?}", message);
+
+            match &message.topic {
+                topic if topic == &self.config.request_topic_list => {
+                    let _success = self
+                        .handle_software_list_request(
+                            mqtt,
+                            plugins.clone(),
+                            &self.config.response_topic_list,
+                            &message,
+                        )
+                        .await
+                        .map_err(|err| {
+                            error!("{:?}", err); // log error and discard such that the agent doesn't exit.
+                        });
+                }
+
+                topic if topic == &self.config.request_topic_update => {
+                    let _success = self
+                        .handle_software_update_request(
+                            mqtt,
+                            plugins.clone(),
+                            &self.config.response_topic_update,
+                            &message,
+                        )
+                        .await
+                        .map_err(|err| {
+                            error!("{:?}", err); // log error and discard such that the agent doesn't exit.
+                        });
+                }
+
+                _ => self.handle_unknown_operation(),
+            }
+        }
+
+        Ok(())
+    }
+
+    fn handle_unknown_operation(&self) {
+        log_error();
+    }
```

Comment:
I would remove this method that is used just once, but I would improve the `log_error` method so it can be use in all places notably in `subscribe_and_process`.

Reply:
We were suppose to decide what to exactly do in the case we receive an unknown operation, this is a placeholder.